File: elixir/1790_are_almost_equal.exs
# https://leetcode.com/problems/check-if-one-string-swap-can-make-strings-equal/
# 05/02/2025

# defmodule Solution do
#   @spec are_almost_equal(s1 :: String.t(), s2 :: String.t()) :: boolean
#   def are_almost_equal(s1, s2) do
#     diff =
#       String.to_charlist(s1)
#       |> Enum.zip(String.to_charlist(s2))
#       |> Enum.filter(fn {c1, c2} -> c1 != c2 end)
#       |> Enum.count()

#     {count1, count2} = {count_chars(s1), count_chars(s2)}

#     (diff == 0 || diff == 2) && count1 == count2
#   end

#   defp count_chars(s) do
#     s
#     |> String.to_charlist()
#     |> Enum.reduce(List.to_tuple(List.duplicate(0, 26)), fn ch, acc ->
#       idx = ch - ?a
#       put_elem(acc, idx, elem(acc, idx) + 1)
#     end)
#   end
# end

# defmodule Solution do
#   @spec are_almost_equal(s1 :: String.t(), s2 :: String.t()) :: boolean
#   def are_almost_equal(s1, s2) when s1 == s2, do: true

#   def are_almost_equal(s1, s2) do
#     case mismatched_pairs(s1, s2) do
#       [{c1, c2}, {c3, c4}] -> c1 == c4 and c2 == c3
#       [] -> true
#       _ -> false
#     end
#   end

#   defp mismatched_pairs(s1, s2) do
#     String.to_charlist(s1)
#     |> Enum.zip(String.to_charlist(s2))
#     |> Enum.filter(fn {c1, c2} -> c1 != c2 end)
#   end
# end

defmodule Solution do
  @spec are_almost_equal(s1 :: String.t(), s2 :: String.t()) :: boolean
  def are_almost_equal(s1, s2) when s1 == s2, do: true

  def are_almost_equal(s1, s2) do
    case mismatched_pairs(s1, s2, []) do
      [{c1, c2}, {c3, c4}] -> c1 == c4 and c2 == c3
      [] -> true
      _ -> false
    end
  end

  defp mismatched_pairs("", "", acc), do: acc

  defp mismatched_pairs(<<c1::utf8, s1::binary>>, <<c2::utf8, s2::binary>>, acc) when c1 != c2,
    do: mismatched_pairs(s1, s2, [{c1, c2} | acc])

  defp mismatched_pairs(<<_c1::utf8, s1::binary>>, <<_c2::utf8, s2::binary>>, acc),
    do: mismatched_pairs(s1, s2, acc)
end

defmodule Main do
  def main do
    test_cases = [
      {
        "bank",
        "kanb",
        true
      },
      {
        "attack",
        "defend",
        false
      },
      {
        "kelb",
        "kelb",
        true
      }
    ]

    test_cases
    |> Enum.each(fn {s1, s2, expected} ->
      result = Solution.are_almost_equal(s1, s2)
      IO.puts("Input: #{inspect({s1, s2})}")
      IO.puts("Output: #{inspect(result)}")
      IO.puts("Expected: #{inspect(expected)}")
      IO.puts("Result: #{if result == expected, do: "PASSED", else: "FAILED"}")
      IO.puts("------")
    end)
  end
end

Main.main()

File: rust/src/bin/1790_are_almost_equal.rs
// https://leetcode.com/problems/check-if-one-string-swap-can-make-strings-equal/
// 05/02/2025

impl Solution {
    pub fn are_almost_equal(s1: String, s2: String) -> bool {
        let diff = s1
            .bytes()
            .zip(s2.bytes())
            .filter(|(c1, c2)| c1 != c2)
            .count();

        let (count1, count2) = (Self::count_char(&s1), Self::count_char(&s2));

        (diff == 0 || diff == 2) && count1 == count2
    }

    #[inline(always)]
    fn count_char(s: &String) -> [i32; 26] {
        s.bytes().fold([0; 26], |mut acc, c| {
            acc[(c - b'a') as usize] += 1;
            acc
        })
    }
}

// impl Solution {
//     pub fn are_almost_equal(s1: String, s2: String) -> bool {
//         if s1 == s2 {
//             return true;
//         }

//         let mismatches: Vec<_> = s1
//             .bytes()
//             .zip(s2.bytes())
//             .enumerate()
//             .filter(|(_, (c1, c2))| c1 != c2)
//             .collect();

//         mismatches.len() == 2
//             && mismatches[0].1 .0 == mismatches[1].1 .1
//             && mismatches[0].1 .1 == mismatches[1].1 .0
//     }
// }

#[derive(Debug)]
struct Solution {}

#[cfg(test)]
mod tests {
    use super::*;
    use rstest::rstest;

    #[rstest]
    #[case("bank".to_string(), "kanb".to_string(), true)]
    #[case("attack".to_string(), "defend".to_string(), false)]
    #[case("kelb".to_string(), "kelb".to_string(), true)]
    #[case("caa".to_string(), "aaz".to_string(), false)]
    fn test(#[case] s1: String, #[case] s2: String, #[case] expected: bool) {
        let result = Solution::are_almost_equal(s1, s2);
        assert_eq!(result, expected);
    }
}

#[cfg(not(test))]
fn main() {
    let test_cases = vec![
        ("bank".to_string(), "kanb".to_string(), true),
        ("attack".to_string(), "defend".to_string(), false),
        ("kelb".to_string(), "kelb".to_string(), true),
    ];

    for (s1, s2, expected) in test_cases {
        let result = Solution::are_almost_equal(s1.clone(), s2.clone());
        println!(
            "Input: {:?}, Expected: {:?}, Got: {:?}",
            (s1, s2),
            expected,
            result
        );
        assert_eq!(result, expected);
    }
}
